diff --git a/cocoa/qcocoafiledialoghelper.mm b/cocoa/qcocoafiledialoghelper.mm
index bb1a9c3..5a4788c 100644
--- a/cocoa/qcocoafiledialoghelper.mm
+++ b/cocoa/qcocoafiledialoghelper.mm
@@ -422,6 +422,13 @@ static QString strippedText(QString s)
     } else {
         QList<QUrl> result;
         QString filename = QString::fromNSString([[mSavePanel URL] path]).normalized(QString::NormalizationForm_C);
+        const QString defaultSuffix = mOptions->defaultSuffix();
+        const QFileInfo fileInfo(filename);
+        // If neither the user or the NSSavePanel have provided a suffix, use
+        // the default suffix (if it exists).
+        if (fileInfo.suffix().isEmpty() && !defaultSuffix.isEmpty()) {
+                filename.append('.').append(defaultSuffix);
+        }
         result << QUrl::fromLocalFile(filename.remove(QLatin1String("___qt_very_unlikely_prefix_")));
         return result;
     }
@@ -449,10 +456,7 @@ static QString strippedText(QString s)
     [mPopUpButton setHidden:chooseDirsOnly];    // TODO hide the whole sunken pane instead?
 
     if (mOptions->acceptMode() == QFileDialogOptions::AcceptSave) {
-        QStringList ext = [self acceptableExtensionsForSave];
-        const QString defaultSuffix = mOptions->defaultSuffix();
-        if (!ext.isEmpty() && !defaultSuffix.isEmpty())
-            ext.prepend(defaultSuffix);
+        const QStringList ext = [self acceptableExtensionsForSave];
         [mSavePanel setAllowedFileTypes:ext.isEmpty() ? nil : qt_mac_QStringListToNSMutableArray(ext)];
     } else {
         [mOpenPanel setAllowedFileTypes:nil]; // delegate panel:shouldEnableURL: does the file filtering for NSOpenPanel
diff --git a/cocoa/qcocoamenu.mm b/cocoa/qcocoamenu.mm
index 77eca5a..14ce632 100644
--- a/cocoa/qcocoamenu.mm
+++ b/cocoa/qcocoamenu.mm
@@ -46,6 +46,7 @@
 #include <QtCore/private/qthread_p.h>
 #include <QtGui/private/qguiapplication_p.h>
 #include "qcocoaapplication.h"
+#include "qcocoaintegration.h"
 #include "qcocoamenuloader.h"
 #include "qcocoamenubar.h"
 #include "qcocoawindow.h"
@@ -596,8 +597,9 @@ void QCocoaMenu::showPopup(const QWindow *parentWindow, const QRect &targetRect,
         [popupCell setMenu:m_nativeMenu];
         [popupCell selectItem:nsItem];
 
-        int availableHeight = screen->availableSize().height();
-        const QPoint &globalPos = parentWindow->mapToGlobal(pos);
+        QCocoaScreen *cocoaScreen = static_cast<QCocoaScreen *>(screen->handle());
+        int availableHeight = cocoaScreen->availableGeometry().height();
+        const QPoint &globalPos = cocoaWindow->mapToGlobal(pos);
         int menuHeight = m_nativeMenu.size.height;
         if (globalPos.y() + menuHeight > availableHeight) {
             // Maybe we need to fix the vertical popup position but we don't know the
